ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,664 @@
+        --------------------------------------------------------------------------------------------------
+        **Object**/ **Condition**   ``allow_multiple_recalls...=True`` ``allow_multiple_recalls...=False``
+        =========================== ================================== ===================================
+        TP                          (C, N\ :sub:`pred`)                (N\ :sub:`pred`,)
+        FP                          (C, N\ :sub:`pred`)                None (FP is computed here to be
+                                                                       faster)
+        scores                      (C, N\ :sub:`pred`)                (N\ :sub:`pred`,)
+        P                           (C,)                               () (A single float)
+        recall                      (C, N\ :sub:`pred`)                (\#unique scores,)
+        precision                   (C, N\ :sub:`pred`)                (\#unique scores,)
+        =========================== ================================== ===================================
+
+        Returns:
+            `(recall, precision)`
+        """
+        indices = torch.argsort(scores, dim=-1, stable=True, descending=True)
+        tp = TP.take_along_dim(indices, dim=-1) if self._task == "classification" else TP[..., indices]
+        tp_summation = tp.cumsum(dim=-1).double()
+        if self._task == "detection" or self.allow_multiple_recalls_at_single_threshold:
+            fp = (
+                cast(torch.Tensor, FP).take_along_dim(indices, dim=-1)
+                if self._task == "classification"
+                else cast(torch.Tensor, FP)[..., indices]
+            )
+            fp_summation = fp.cumsum(dim=-1).double()
+        if not self.allow_multiple_recalls_at_single_threshold:
+            # Adopted from Scikit-learn's implementation
+            unique_scores_indices = torch.nonzero(
+                scores.take_along_dim(indices).diff(append=(scores.max() + 1).unsqueeze(dim=0)), as_tuple=True
+            )[0]
+            tp_summation = tp_summation[..., unique_scores_indices]
+            if self._task == "classification":
+                fp_summation = (unique_scores_indices + 1) - tp_summation
+            else:
+                fp_summation = fp_summation[..., unique_scores_indices]
+
+        if self._task == "classification" and self.allow_multiple_recalls_at_single_threshold:
+            recall = torch.where(P == 0, 1, tp_summation.T / P).T
+        elif self._task == "classification" and P == 0:
+            recall = torch.ones_like(tp_summation, device=self._device, dtype=torch.bool)
+        else:
+            recall = tp_summation / P
+        # precision = tp_summation / (fp_summation + tp_summation + torch.finfo(torch.double).eps)
+        # or
+        predicted_positive = tp_summation + fp_summation
+        precision = tp_summation / torch.where(predicted_positive == 0, 1, predicted_positive)
+        return recall, precision
```

Comment:
Same comment here as well, I would prefer to dispatch the code by task with inheritance than with if/else.